bestie/utils/googlebench/bestie_reporter.cc

```diff
@@ -0,0 +1,180 @@
+#include "bestie/utils/googlebench/bestie_reporter.hh"
+
+#include <google/protobuf/text_format.h>
+
+#include <chrono>
+#include <fstream>
+
+#include "google/protobuf/io/zero_copy_stream.h"
+#include "google/protobuf/io/zero_copy_stream_impl.h"
+
+namespace bestie {
+
+std::string MetricsPath(const std::string_view filename, std::string_view path) {
+  return absl::StrCat(path, "/", filename, ".metrics.pb");
+}
+
+std::string MetricsPath(const std::string_view filename) {
+  // getenv is not thread safe, static variable initialization is.
+  // Also, try to discourage changing the environment variables at run time.
+  static const char* path = []() {
+    if (const char* path = std::getenv("TEST_UNDECLARED_OUTPUTS_DIR"); path) return path;
+    if (const char* path = std::getenv("TMPDIR"); path) return path;
+
+    return "/tmp";
+  }();
+
+  return MetricsPath(filename, path);
+}
+
+std::string MetricsPath(int attempt, const std::string_view filename) {
+  if (attempt <= 0) return MetricsPath(filename);
+  return MetricsPath(absl::StrFormat("%s%03d", filename, attempt));
+}
+
+static void AddTag(bestie::proto::TestMetric* metric, const std::string& key,
+                   const std::string& value) {
+  auto* tag = metric->add_tags();
+  tag->set_key(key);
+  tag->set_value(value);
+}
+
+static bestie::proto::TestMetric* AddMetric(bestie::proto::TestMetrics* metrics,
+                                            const std::string& name, double value,
+                                            const auto&... prototype) {
+  auto* metric = metrics->add_metrics();
+
+  metric->set_metricname(name);
+  metric->set_value(value);
+
+  (metric->MergeFrom(prototype), ...);
+
+  return metric;
+}
+
+bool Reporter::ReportContext(const Context& context) {
+  int64_t epoch = std::chrono::duration_cast<std::chrono::nanoseconds>(
+                      std::chrono::system_clock::now().time_since_epoch())
+                      .count();
+
+  context_.Clear();
+  context_.set_timestamp(epoch);
+
+  AddTag(&context_, "context__sys_info__name", context.sys_info.name);
+  AddTag(&context_, "context__executable_name", context.executable_name);
```

Comment:
Is context__executable_name just the executable name with the path stripped, or does it contain absolute or relative path?